File: src/lae/daemon/context_sync.py
"""Sync taskspace from the active Hyprland workspace name."""

from __future__ import annotations

from lae.core.models import ContextMode, SessionState
from lae.core.workspaces import (
    allowed_workspace_names,
    is_default_taskspace_workspace_name,
    task_for_workspace_name,
)
from lae.integrations import hyprland


def sync_from_active_workspace(state: SessionState) -> bool:
    """Align lae taskspace + active workspace index with Hyprland focus."""
    if not hyprland.available():
        return False

    active = hyprland.get_active_workspace()
    if not active or not active.name:
        return False

    changed = False
    name = active.name

    if state.context_mode != ContextMode.global_:
        if is_default_taskspace_workspace_name(name, state.default_workspace_count):
            if state.context_mode != ContextMode.default or state.current_task_id is not None:
                state.context_mode = ContextMode.default
                state.current_task_id = None
                changed = True
        else:
            task = task_for_workspace_name(state, name)
            if task:
                if (
                    state.context_mode != ContextMode.task
                    or state.current_task_id != task.id
                ):
                    state.context_mode = ContextMode.task
                    state.current_task_id = task.id
                    changed = True

    names = allowed_workspace_names(
        state, workspace_count=state.default_workspace_count
    )
    if name in names:
        rel = names.index(name) + 1
        key = state.taskspace_key()
        if state.last_workspace.get(key) != rel:
            state.last_workspace[key] = rel
            changed = True

    return changed

File: crates/lae-core/src/context_sync.rs
use crate::models::SessionState;
use crate::workspaces::{
    allowed_workspace_names, is_default_taskspace_workspace_name, task_for_workspace_name,
};

/// Align taskspace with a known workspace name (no Hyprland IPC).
pub fn sync_from_workspace_name(state: &mut SessionState, name: &str) -> bool {
    if name.is_empty() {
        return false;
    }

    let mut changed = false;

    if is_default_taskspace_workspace_name(name, state.default_workspace_count) {
        if state.context_mode != crate::models::ContextMode::Default
            || state.current_task_id.is_some()
        {
            state.context_mode = crate::models::ContextMode::Default;
            state.current_task_id = None;
            changed = true;
        }
    } else if let Some(task) = task_for_workspace_name(state, name) {
        let task_id = task.id.clone();
        if state.context_mode != crate::models::ContextMode::Task
            || state.current_task_id.as_deref() != Some(task_id.as_str())
        {
            state.context_mode = crate::models::ContextMode::Task;
            state.current_task_id = Some(task_id);
            changed = true;
        }
    }

    let names = allowed_workspace_names(state);
    if let Some(idx) = names.iter().position(|n| n == name) {
        let rel = (idx + 1) as i32;
        let key = state.taskspace_key();
        if state.last_workspace.get(&key).copied() != Some(rel) {
            state.last_workspace.insert(key, rel);
            changed = true;
        }
    }

    changed
}

pub fn sync_from_active_workspace(state: &mut SessionState) -> bool {
    if !crate::hyprland::available() {
        return false;
    }

    let Ok(Some(active)) = crate::hyprland::get_active_workspace() else {
        return false;
    };
    if active.name.is_empty() {
        return false;
    }

    sync_from_workspace_name(state, &active.name)
}
